In [1]:
import pandas as pd
import numpy as np
import seaborn as sns

df = sns.load_dataset('titanic')

.() function used to read the top few rows in a data group

In [5]:
df.head() 

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0         0       3    male  22.0      1      0   7.2500        S   Third   
1         1       1  female  38.0      1      0  71.2833        C   First   
2         1       3  female  26.0      0      0   7.9250        S   Third   
3         1       1  female  35.0      1      0  53.1000        S   First   
4         0       3    male  35.0      0      0   8.0500        S   Third   
5         0       3    male   NaN      0      0   8.4583        Q   Third   
6         0       1    male  54.0      0      0  51.8625        S   First   
7         0       3    male   2.0      3      1  21.0750        S   Third   
8         1       3  female  27.0      0      2  11.1333        S   Third   
9         1       2  female  14.0      1      0  30.0708        C  Second   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True  
5    man        True  NaN   Queenstown    no   True  
6    man        True    E  Southampton    no   True  
7  child       False  NaN  Southampton    no  False  
8  woman       False  NaN  Southampton   yes  False  
9  child       False  NaN    Cherbourg   yes  False

describe() function used to get some general info on the data set

In [6]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

loop to print the .unique() results in each column, to see the variety of values, and especially null values

In [3]:
for col in df:
  print(df[col].unique())

[0 1]
[3 1 2]
['male' 'female']
[22.   38.   26.   35.     nan 54.    2.   27.   14.    4.   58.   20.
 39.   55.   31.   34.   15.   28.    8.   19.   40.   66.   42.   21.
 18.    3.    7.   49.   29.   65.   28.5   5.   11.   45.   17.   32.
 16.   25.    0.83 30.   33.   23.   24.   46.   59.   71.   37.   47.
 14.5  70.5  32.5  12.    9.   36.5  51.   55.5  40.5  44.    1.   61.
 56.   50.   36.   45.5  20.5  62.   41.   52.   63.   23.5   0.92 43.
 60.   10.   64.   13.   48.    0.75 53.   57.   80.   70.   24.5   6.
  0.67 30.5   0.42 34.5  74.  ]
[1 0 3 4 2 5 8]
[0 1 2 5 3 4 6]
[  7.25    71.2833   7.925   53.1      8.05     8.4583  51.8625  21.075
  11.1333  30.0708  16.7     26.55    31.275    7.8542  16.      29.125
  13.      18.       7.225   26.       8.0292  35.5     31.3875 263.
   7.8792   7.8958  27.7208 146.5208   7.75    10.5     82.1708  52.
   7.2292  11.2417   9.475   21.      41.5792  15.5     21.6792  17.8
  39.6875   7.8     76.7292  61.9792  27.75    46.9    

dropna() drops all the Null values

In [29]:
new_df = df.dropna()
new_df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
1          1       1  female  38.0      1      0  71.2833        C  First   
3          1       1  female  35.0      1      0  53.1000        S  First   
6          0       1    male  54.0      0      0  51.8625        S  First   
10         1       3  female   4.0      1      1  16.7000        S  Third   
11         1       1  female  58.0      0      0  26.5500        S  First   

      who  adult_male deck  embark_town alive  alone  
1   woman       False    C    Cherbourg   yes  False  
3   woman       False    C  Southampton   yes  False  
6     man        True    E  Southampton    no   True  
10  child       False    G  Southampton   yes  False  
11  woman       False    C  Southampton   yes   True

However, we can see that the count has dropped from 891 to 182, so is too severe...

In [30]:
new_df.count()

survived       182
pclass         182
sex            182
age            182
sibsp          182
parch          182
fare           182
embarked       182
class          182
who            182
adult_male     182
deck           182
embark_town    182
alive          182
alone          182
dtype: int64

duplicated() method checks for duplicated rows

In [4]:
df[df. duplicated()]

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
47          1       3  female   NaN      0      0   7.7500        Q   Third   
76          0       3    male   NaN      0      0   7.8958        S   Third   
77          0       3    male   NaN      0      0   8.0500        S   Third   
87          0       3    male   NaN      0      0   8.0500        S   Third   
95          0       3    male   NaN      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
870         0       3    male  26.0      0      0   7.8958        S   Third   
877         0       3    male  19.0      0      0   7.8958        S   Third   
878         0       3    male   NaN      0      0   7.8958        S   Third   
884         0       3    male  25.0      0      0   7.0500        S   Third   
886         0       2    male  27.0      0      0  13.0000        S  Second   

       who  adult_male deck  embark_town alive  alone  
47   woman       False  NaN   Queenstown   yes   True  
76     man        True  NaN  Southampton    no   True  
77     man        True  NaN  Southampton    no   True  
87     man        True  NaN  Southampton    no   True  
95     man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
870    man        True  NaN  Southampton    no   True  
877    man        True  NaN  Southampton    no   True  
878    man        True  NaN  Southampton    no   True  
884    man        True  NaN  Southampton    no   True  
886    man        True  NaN  Southampton    no   True  

[107 rows x 15 columns]

then, drop_duplicates() is used to take these out
(inplace=True) here specifies that this change should be made permanently to the original DataFrame, rather than in a copy

In [62]:
df.drop_duplicates(inplace = True)

groupby() method is used to group by a column, as in SQL
mean() to generate an average

In [66]:
df.groupby('class')['fare', 'age'].mean().round()

/var/folders/3r/xcjg5vss2p7ckpl_ks5qx9gm0000gn/T/ipykernel_78108/1373306208.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  df.groupby('class')['fare', 'age'].mean().round()


fare   age
class             
First   84.0  38.0
Second  22.0  30.0
Third   14.0  25.0

len() returns values based on column labels (like SELECT in SQL)... here we can return all the rows where sex = female

In [57]:
len(df[df['sex']=='female'])

293

adding this method into a new dictionary, we can then create a new DataFrame with the data

In [58]:
women_and_men = {
    "total women": [len(df[df['sex']=='female'])],
    "total men": [len(df[df['sex']=='male'])]
}
df2 = pd.DataFrame(women_and_men)
df2

total women  total men
0          293        491

we can return the total fare using the .sum() method
(skipna=True) here specifies that all Null values should be ignored

In [59]:
total_fare = df['fare'].sum(skipna=True)
total_fare

27214.003900000003

we can then use the .apply() method to add in another column, in which we caluclate the percentage of total fare using the total_fare variable from above

In [65]:
df['Percentage of Total Fare'] = df.apply(lambda row: (row['fare']/total_fare * 100).round(4), axis=1)
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  Percentage of Total Fare  
0    man        True  NaN  Southampton    no  False                      0.03  
1  woman       False    C    Cherbourg   yes  False                      0.26  
2  woman       False  NaN  Southampton   yes   True                      0.03  
3  woman       False    C  Southampton   yes  False                      0.20  
4    man        True  NaN  Southampton    no   True                      0.03